aws/request/timeout_read_closer.go

```diff
@@ -0,0 +1,78 @@
+package request
+
+import (
+	"io"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+var timeoutErr = awserr.New(
+	ErrCodeResponseTimeout,
+	"read on body has reached the timeout limit",
+	nil,
+)
+
+type readResult struct {
+	n   int
+	err error
+}
+
+// timeoutReadCloser will handle body reads that take too long.
+// We will return a ErrReadTimeout error if a timeout occurs.
+type timeoutReadCloser struct {
+	reader   io.ReadCloser
+	duration time.Duration
+}
+
+// Read will spin off a goroutine to call the reader's Read method. We will
+// select on the timer's channel or the read's channel. Whoever completes first
+// will be returned.
+func (r *timeoutReadCloser) Read(b []byte) (int, error) {
+	timer := time.NewTimer(r.duration)
+	c := make(chan readResult, 1)
+
+	go func() {
+		n, err := r.reader.Read(b)
+		timer.Stop()
```

Comment:
Forgive me if this is what is intended, but wouldn't stopping the timer here essentially be putting a timeout on the **first** successful `Read()` call on the HTTPResponse.Body?

This could still lead to a situation where the Kinesis service stops sending data, improperly hangs up the connection, or if the TCP FIN was lost in the network. 

As before, a global `http.Client{Timeout:xxx}` could catch this problem, but that is applied across all clients in use by a particular session.